pkg/testing/ig/ig.go

```diff
@@ -0,0 +1,244 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ig
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"syscall"
+
+	"github.com/blang/semver"
+)
+
+type IG struct {
+	path    string
+	image   string
+	version semver.Version
+
+	cmd    *exec.Cmd
+	env    []string
+	flags  []string
+	Stdout bytes.Buffer
+	Stderr bytes.Buffer
+
+	// StartAndStop indicates this command should first be started then stopped.
+	// It corresponds to gadget like execsnoop which wait user to type Ctrl^C.
+	startAndStop bool
+
+	// started indicates this command was started.
+	// It is only used by command which have StartAndStop set.
+	started bool
+}
+
+func (ig *IG) IsStartAndStop() bool {
+	return ig.startAndStop
+}
+
+func (ig *IG) Running() bool {
+	return ig.started
+}
+
+func (ig *IG) createCmd() {
+	args := append([]string{"run", ig.image}, ig.flags...)
+	cmd := exec.Command(ig.path, args...)
+
+	ig.env = append(ig.env, "IG_EXPERIMENTAL=true")
+	cmd.Env = append(cmd.Env, ig.env...)
+
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = &ig.Stdout
+	cmd.Stderr = &ig.Stderr
+
+	// To be able to kill the process of ig.cmd,
+	// we need to send the termination signal to their process group ID
+	// (PGID). However, child processes get the same PGID as their parents by
+	// default, so in order to avoid killing also the integration tests process,
+	// we set the fields Setpgid and Pgid of syscall.SysProcAttr before
+	// executing ig.cmd. Doing so, the PGID of ig.cmd (and its children)
+	// will be set to its process ID, see:
+	// https://cs.opensource.google/go/go/+/refs/tags/go1.17.8:src/syscall/exec_linux.go;l=32-34.
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true, Pgid: 0}
+
+	ig.cmd = cmd
+}
+
+func (ig *IG) RunCmd() error {
+	ig.createCmd()
+	if err := ig.cmd.Run(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ig *IG) StartCmd() error {
+	if ig.started {
+		return fmt.Errorf("Warn(%v): trying to start command but it was already started", ig.cmd)
+	}
+
+	ig.createCmd()
+	if err := ig.cmd.Start(); err != nil {
+		return err
+	}
+
+	ig.started = true
+
+	return nil
+}
+
+// kill kills a command by sending SIGKILL because we want to stop the process
+// immediatly and avoid that the signal is trapped.
+func (ig *IG) kill() error {
+	const sig syscall.Signal = syscall.SIGKILL
+
+	// No need to kill, command has not been executed yet or it already exited
+	if ig.cmd == nil || (ig.cmd.ProcessState != nil && ig.cmd.ProcessState.Exited()) {
+		return nil
+	}
+
+	// Given that we set Setpgid, here we just need to send the PID of ig.cmd
+	// (which is the same PGID) as a negative number to syscall.Kill(). As a
+	// result, the signal will be received by all the processes with such PGID,
+	// in our case, the process of ig.cmd.
+	err := syscall.Kill(-ig.cmd.Process.Pid, sig)
```

Comment:
Things were updated to make `igrunner` a wrapper of `command` so that it reuses all the `command`'s logic for killing the process, which is generic for any command.